step-generation/src/getNextRobotStateAndWarnings/stackerUpdates.ts

```diff
@@ -0,0 +1,196 @@
+  const { moduleId, count } = params
+  const moduleState = _getStackerModuleState(robotState, moduleId)
+
+  if (moduleState != null) {
+    if (count != null && count > 0) {
+      moduleState.labwareIdsInStacker =
+        moduleState?.labwareIdsInStacker?.splice(
+          moduleState?.labwareIdsInStacker?.length - 1 - count
+        ) ?? null
+    } else {
+      moduleState.labwareIdsInStacker = null
+    }
+  }
+}
+
+export const forFlexStackerFill = (
+  params: FlexStackerFillParams,
+  invariantContext: InvariantContext,
+  robotStateAndWarnings: RobotStateAndWarnings
+): void => {
+  const { robotState } = robotStateAndWarnings
+  const { moduleId, count } = params
+  const moduleState = _getStackerModuleState(robotState, moduleId)
+  const labwareDefinition =
+    invariantContext.labwareEntities[
+      moduleState?.labwareIdsInStacker?.[0] ?? ''
+    ]?.def
+  const listOfLabwareDefinitions = Array.from(
+    { length: moduleState?.labwareIdsInStacker?.length ?? 0 },
+    _ => labwareDefinition
+  )
+  const poolHeight = getHeightOfLabwareStackFromDefinitions(
+    listOfLabwareDefinitions
+  )
+  const poolOverlap = getLabwareOverlapOffset(
+    FLEX_STACKER_MODULE_V1,
+    labwareDefinition,
+    'default'
+  )
+  const maxStorableLabware = getStackerMaxPoolCountByHeight(
+    FLEX_STACKER_MODULE_V1,
+    poolHeight,
+    poolOverlap.z
+  )
+
+  if (moduleState != null) {
+    if (
+      count != null &&
+      count > 0 &&
+      maxStorableLabware >
+        count + (moduleState.labwareIdsInStacker?.length ?? 0)
+    ) {
+      // create labware entities for the new labware
+      const newLabwareIdList = Array.from({ length: count }, () => uuid())
+      moduleState.labwareIdsInStacker = [
+        ...(moduleState.labwareIdsInStacker ?? []),
+        ...newLabwareIdList,
+      ]
+    }
+  }
+}
+
+export const forFlexStackerRetrieve = (
+  params: ModuleOnlyParams,
+  invariantContext: InvariantContext,
+  robotStateAndWarnings: RobotStateAndWarnings
+): void => {
+  const { robotState } = robotStateAndWarnings
+  const { moduleId } = params
+  const moduleState = _getStackerModuleState(robotState, moduleId)
+  if (moduleState != null) {
+    if (moduleState.shuttlePosition === 'retrieved') {
+      throw new Error(
+        'Cannot retrieve labware bc there is labware on the shuttle'
+      )
+    }
+    if (moduleState.labwareIdsInStacker?.length === 0) {
+      throw new Error(
+        'Cannot retrieve labware bc there is no labware in the stacker'
+      )
+    }
+    if (moduleState.storedLabwareDetails?.primaryLabware == null)
+      throw new Error(
+        'Cannot retrieve labware bc there is no stored labware details or primary labware'
+      )
+    moduleState.shuttlePosition = 'retrieved'
+  }
+
+  const retrievedLabware = moduleState?.labwareIdsInStacker?.shift()
+  if (retrievedLabware == null) {
+    throw new Error(
+      'Cannot retrieve labware bc there is no labware in the stacker'
+    )
+  }
```

Comment:
can you update all of these to `consol.errors`? 
```suggestion
    if (moduleState.shuttlePosition === 'retrieved') {
      console.error(
        'Cannot retrieve labware bc there is labware on the shuttle'
      )
    }
    if (moduleState.labwareIdsInStacker?.length === 0) {
      console.error(
        'Cannot retrieve labware bc there is no labware in the stacker'
      )
    }
    if (moduleState.storedLabwareDetails?.primaryLabware == null)
      console.error(
        'Cannot retrieve labware bc there is no stored labware details or primary labware'
      )
    moduleState.shuttlePosition = 'retrieved'
  }

  const retrievedLabware = moduleState?.labwareIdsInStacker?.shift()
  if (retrievedLabware == null) {
    console.error(
      'Cannot retrieve labware bc there is no labware in the stacker'
    )
  }
```